src/order/heyting/boundary.lean

```diff
@@ -0,0 +1,113 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import order.boolean_algebra
+
+/-!
+# Co-Heyting boundary
+
+The boundary of an element of a co-Heyting algebra is the intersection of its Heyting negation with
+itself. The boundary in the co-Heyting algebra of closed sets coincides with the topological
+boundary.
+
+## Main declarations
+
+* `coheyting.boundary`: Co-Heyting boundary. `coheyting.boundary a = a ⊓ ￢a`
+
+## Notation
+
+`∂ a` is notation for `coheyting.boundary a` in locale `heyting`.
+-/
+
+variables {α : Type*}
+
+namespace coheyting
+variables [coheyting_algebra α] {a b : α}
+
+/-- The boundary of an element of a co-Heyting algebra is the intersection of its Heyting negation
+with itself. Note that this is always `⊥` for a boolean algebra. -/
+def boundary (a : α) : α := a ⊓ ￢a
+
+localized "prefix `∂ `:120 := coheyting.boundary" in heyting
+
+lemma inf_hnot_self (a : α) : a ⊓ ￢a = ∂ a := rfl
+
+lemma boundary_le : ∂ a ≤ a := inf_le_left
+lemma boundary_le_hnot : ∂ a ≤ ￢a := inf_le_right
+
+@[simp] lemma boundary_bot : ∂ (⊥ : α) = ⊥ := bot_inf_eq
+@[simp] lemma boundary_top : ∂ (⊤ : α) = ⊥ := by rw [boundary, hnot_top, inf_bot_eq]
+
+lemma boundary_hnot_le (a : α) : ∂ ￢a ≤ ∂ a := inf_comm.trans_le $ inf_le_inf_right _ hnot_hnot_le
+
+@[simp] lemma boundary_hnot_hnot (a : α) : ∂ ￢￢a = ∂ ￢a :=
+by simp_rw [boundary, hnot_hnot_hnot, inf_comm]
+
+lemma hnot_boundary (a : α) : ￢ ∂ a = ￢a ⊔ ￢￢a := hnot_inf_distrib _ _
```

Comment:
Don't we even have that:
```suggestion
lemma hnot_boundary (a : α) : ￢ ∂ a = ⊤ := sorry
```
because, dually, this is `¬ a ∧ ¬¬ a` which is exactly `⊥`.

Reply:
In fact, [nlab claims that](https://ncatlab.org/nlab/show/co-Heyting+boundary) `∃ a, x = ∂ a ↔ ¬ x = 1`!